### ISSUE TITLE
Fails to compile on FreeBSD 13-CURRENT (meson)

### ISSUE THREAD

**Issue by diizzyy** [p11-glue/p11-kit#346]
On very recent builds of FreeBSD 13-CURRENT the build fails using 0.23.22 and commit 507c394cfcf4edffc5e4450c5d737e545c26b857

```
The Meson build system
Version: 0.56.0
Source dir: /usr/ports/security/p11-kit/work/p11-kit-0.23.22
Build dir: /usr/ports/security/p11-kit/work/p11-kit-0.23.22/_build
Build type: native build
Project name: p11-kit
Project version: 0.23.22
Using 'CC' from environment with value: 'cc'
Using 'CFLAGS' from environment with value: '-O2 -pipe -march=ivybridge  -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing '
Using 'LDFLAGS' from environment with value: ' -fstack-protector-strong '
Using 'CPPFLAGS' from environment with value: '-isystem /usr/local/include'
C compiler for the host machine: cc (clang 11.0.0 "FreeBSD clang version 11.0.0 (git@github.com:llvm/llvm-project.git llvmorg-11.0.0-0-g176249bd673)")
C linker for the host machine: cc ld.lld 11.0.0
Using 'CC' from environment with value: 'cc'
Using 'CFLAGS' from environment with value: '-O2 -pipe -march=ivybridge  -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing '
Using 'LDFLAGS' from environment with value: ' -fstack-protector-strong '
Using 'CPPFLAGS' from environment with value: '-isystem /usr/local/include'
Host machine cpu family: x86_64
Host machine cpu: x86_64
Checking for size of "unsigned long" : 8
Run-time dependency threads found: YES
Checking for function "pthread_create" with dependency threads: YES
Checking for function "nanosleep" : YES
Checking for function "dlopen" : YES
Library nsl found: NO
Has header "locale.h" : YES
Checking for type "locale_t" : YES
Checking for function "newlocale" : YES
Checking for function "strerror_l" : YES
Has header "sys/resource.h" : YES
Has header "ucred.h" : NO
Checking for function "fdwalk" : NO
Checking for function "getauxval" : NO
Checking for function "getexecname" : NO
Checking for function "getpeereid" : YES
Checking for function "getpeerucred" : NO
Checking for function "getprogname" : YES
Checking for function "getresuid" : YES
Checking for function "issetugid" : YES
Checking for function "mkdtemp" : YES
Checking for function "mkstemp" : YES
Checking for function "secure_getenv" : NO
Checking for function "strndup" : YES
Checking whether type "struct dirent" has member "d_type" : YES
Checking if "thread-local storage class" compiles: YES
Checking for function "gmtime_r" : YES
Checking if "program_invocation_short_name_test_code" links: NO
Checking if "__progname" links: YES
Checking if "__libc_enable_secure" links: NO
Checking if "vsock_test" compiles: NO
Checking for type "sighandler_t" : NO
Checking for type "sig_t" : NO
Checking for type "__sighandler_t" : NO
Checking for type "sighandler_t" : NO
Checking for type "sig_t" : YES
Checking for type "__sighandler_t" : YES
Has header "stdbool.h" : YES
Checking for function "asprintf" : YES
Checking for function "basename" : YES
Checking for function "memdup" : NO
Checking for function "reallocarray" : YES
Checking for function "secure_getenv" : NO (cached)
Checking for function "setenv" : YES
Checking for function "strerror_r" : YES
Checking for function "strnstr" : YES
Checking for function "vasprintf" : YES
Header <errno.h> has symbol "program_invocation_short_name" : NO
Header <stdio.h> has symbol "asprintf" : YES
Header <stdio.h> has symbol "vasprintf" : YES
Header <stdlib.h> has symbol "reallocarray" : YES
Using 'PKG_CONFIG' from environment with value: 'pkgconf'
Found pkg-config: /usr/local/bin/pkgconf (1.7.3)
Run-time dependency libffi found: YES 3.3
Run-time dependency libtasn1 found: YES 4.16.0
Program asn1Parser found: YES (/usr/local/bin/asn1Parser)
Did not find CMake 'cmake'
Found CMake: NO
Run-time dependency libsystemd found: NO (tried pkgconfig and cmake)
Run-time dependency systemd found: NO (tried pkgconfig and cmake)
Configuring config.h using configuration
Compiler for C supports link arguments -Wl,--version-script,/usr/ports/security/p11-kit/work/p11-kit-0.23.22/p11-kit/libp11-kit.map: YES
Program meson_post_install.sh found: YES (/usr/ports/security/p11-kit/work/p11-kit-0.23.22/p11-kit/meson_post_install.sh)
Compiler for C supports link arguments -Wl,--version-script,/usr/ports/security/p11-kit/work/p11-kit-0.23.22/p11-kit/p11-module.map: YES
Program gen-pkcs11-gnu.sh found: YES (/usr/ports/security/p11-kit/work/p11-kit-0.23.22/p11-kit/gen-pkcs11-gnu.sh)
Program test-messages.sh found: YES (/usr/ports/security/p11-kit/work/p11-kit-0.23.22/p11-kit/test-messages.sh)
Program test-server.sh found: YES (/usr/ports/security/p11-kit/work/p11-kit-0.23.22/p11-kit/test-server.sh)
Configuring pkcs11.conf.example using configuration
Run-time dependency glib-2.0 found: YES 2.66.3
Program gtkdoc-scan found: YES (/usr/local/bin/gtkdoc-scan)
Program gtkdoc-scangobj found: YES (/usr/local/bin/gtkdoc-scangobj)
Program gtkdoc-mkdb found: YES (/usr/local/bin/gtkdoc-mkdb)
Program gtkdoc-mkhtml found: YES (/usr/local/bin/gtkdoc-mkhtml)
Program gtkdoc-fixxref found: YES (/usr/local/bin/gtkdoc-fixxref)
dependency <CustomTarget 3b82860@@generate version.xml@cus: ['echo', '-n', '0.23.22']> not handled to build gir files
dependency <CustomTarget 3b82860@@generate userdir.xml@cus: ['echo', '-n', '~/.config/pkcs11']> not handled to build gir files
dependency <CustomTarget 3b82860@@generate sysdir.xml@cus: ['echo', '-n', 'etc/pkcs11']> not handled to build gir files
Program xsltproc found: YES (/usr/local/bin/xsltproc)
Run-time dependency bash-completion found: YES 2.11
Build targets in project: 106

Found ninja-1.10.2 at /usr/local/bin/ninja
===>  Building for p11-kit-0.23.22
[1/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/array.c.o -MF common/libp11-common.a.p/array.c.o.d -o common/libp11-common.a.p/array.c.o -c ../common/array.c
[2/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/argv.c.o -MF common/libp11-common.a.p/argv.c.o.d -o common/libp11-common.a.p/argv.c.o -c ../common/argv.c
[3/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/buffer.c.o -MF common/libp11-common.a.p/buffer.c.o.d -o common/libp11-common.a.p/buffer.c.o -c ../common/buffer.c
[4/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/compat.c.o -MF common/libp11-common.a.p/compat.c.o.d -o common/libp11-common.a.p/compat.c.o -c ../common/compat.c
../common/compat.c:856:7: warning: implicit declaration of function 'getresuid' is invalid in C99 [-Wimplicit-function-declaration]
                if (getresuid (&ruid, &euid, &suid) != 0 ||
                    ^
../common/compat.c:857:7: warning: implicit declaration of function 'getresgid' is invalid in C99 [-Wimplicit-function-declaration]
                    getresgid (&rgid, &egid, &sgid) != 0)
                    ^
../common/compat.c:942:7: warning: unused variable 'num' [-Wunused-variable]
        long num;
             ^
3 warnings generated.
[5/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/constants.c.o -MF common/libp11-common.a.p/constants.c.o.d -o common/libp11-common.a.p/constants.c.o -c ../common/constants.c
[6/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/message.c.o -MF common/libp11-common.a.p/message.c.o.d -o common/libp11-common.a.p/message.c.o -c ../common/message.c
FAILED: common/libp11-common.a.p/message.c.o
cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/message.c.o -MF common/libp11-common.a.p/message.c.o.d -o common/libp11-common.a.p/message.c.o -c ../common/message.c
../common/message.c:64:1: error: unknown type name 'locale_t'
locale_t p11_message_locale = (locale_t) 0;
^
../common/message.c:64:32: error: use of undeclared identifier 'locale_t'
locale_t p11_message_locale = (locale_t) 0;
                               ^
../common/message.c:123:29: error: use of undeclared identifier 'locale_t'
        if (p11_message_locale != (locale_t) 0)
                                   ^
../common/message.c:124:20: warning: implicit declaration of function 'strerror_l' is invalid in C99 [-Wimplicit-function-declaration]
                strncpy (strerr, strerror_l (errnum, p11_message_locale), sizeof (strerr));
                                 ^
1 warning and 3 errors generated.
[7/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/hash.c.o -MF common/libp11-common.a.p/hash.c.o.d -o common/libp11-common.a.p/hash.c.o -c ../common/hash.c
[8/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/debug.c.o -MF common/libp11-common.a.p/debug.c.o.d -o common/libp11-common.a.p/debug.c.o -c ../common/debug.c
[9/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/lexer.c.o -MF common/libp11-common.a.p/lexer.c.o.d -o common/libp11-common.a.p/lexer.c.o -c ../common/lexer.c
[10/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/dict.c.o -MF common/libp11-common.a.p/dict.c.o.d -o common/libp11-common.a.p/dict.c.o -c ../common/dict.c
[11/286] cc -Icommon/libp11-common.a.p -Icommon -I../common -I. -I.. -fvisibility=hidden -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O3 -D_GNU_SOURCE -DP11_KIT_FUTURE_UNSTABLE_API -O2 -march=ivybridge -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing -isystem /usr/local/include -fPIC -MD -MQ common/libp11-common.a.p/attrs.c.o -MF common/libp11-common.a.p/attrs.c.o.d -o common/libp11-common.a.p/attrs.c.o -c ../common/attrs.c
ninja: build stopped: subcommand failed.
```

Adding ```#include <xlocale.h>``` to https://github.com/p11-glue/p11-kit/blob/master/common/message.c#L54 (below this line) fixes the issue although I'm not sure this is the correct way of fixing it however unit tests completes.

**Comment by ueno** [comment]
I need to set up another VM for testing, but does it work with autotools?

It seems strange because `locale_t` is defined in `<locale.h>` in POSIX and `messages.c` has `_POSIX_C_SOURCE` set to `200112L`.

**Comment by diizzyy** [comment]
Strange indeed, autotools also fails in similar fasion

**Comment by diizzyy** [comment]
It has come to my attention that default locate changed to ```C.UTF-8``` which might be causing the issues?

**Comment by diizzyy** [comment]
This seems to have fixed itself, can't reproduce failure in 63f93c7e11e3
